dashboard/modules/runtime_env/runtime_env_agent.py

```diff
@@ -102,17 +123,50 @@ def run_setup_with_logger():
                 per_job_logger.debug(f"Worker has resource :"
                                      f"{allocated_resource}")
                 context = RuntimeEnvContext(env_vars=runtime_env.env_vars())
-                self._pip_manager.setup(
-                    runtime_env, context, logger=per_job_logger)
-                self._conda_manager.setup(
-                    runtime_env, context, logger=per_job_logger)
-                self._py_modules_manager.setup(
-                    runtime_env, context, logger=per_job_logger)
-                self._working_dir_manager.setup(
-                    runtime_env, context, logger=per_job_logger)
                 self._container_manager.setup(
                     runtime_env, context, logger=per_job_logger)
 
+                for (manager, uri_cache) in [(self._working_dir_manager,
+                                              self._working_dir_uri_cache),
+                                             (self._conda_manager,
+                                              self._conda_uri_cache),
+                                             (self._pip_manager,
+                                              self._pip_uri_cache)]:
+                    uri = manager.get_uri(runtime_env)
+                    if uri is not None:
+                        if uri not in uri_cache:
+                            size_bytes = manager.create(
+                                uri,
+                                runtime_env,
+                                context,
+                                logger=per_job_logger)
+                            uri_cache.add(
+                                uri, size_bytes, logger=per_job_logger)
+                        else:
+                            uri_cache.mark_used(uri, logger=per_job_logger)
+                    manager.modify_context(uri, runtime_env, context)
+
+                # Set up py_modules. For now, py_modules uses multiple URIs so
+                # the logic is slightly different from working_dir, conda, and
+                # pip above.
+                py_modules_uris = self._py_modules_manager.get_uris(
```

Comment:
what is the plan for unifying this>?

Reply:
In a followup we should use a single URI in the agent.  The py_modules plugin itself will manage the "sub-uris" representing individual modules, but the agent won't know about this.  (This is the same strategy we had in mind for the Anyscale proprietary working_dir, Proposal 2 here: https://docs.google.com/document/d/18X8rx6Q-vYj6FyYWQEjqqnH0KRx97-OrTDuw7NwV5PM/edit#heading=h.pocjsgysu6xf )

Reply:
In a followup we should use a single URI in the agent. The py_modules plugin itself will manage the "sub-uris" representing individual modules, but the agent won't know about this.  (This is similar to what we had discussed earlier for a more complicated working_dir plugin using multiple URIs)

Reply:
this makes sense